iroh-net/src/client.rs

```diff
@@ -5,129 +5,37 @@ use std::{
-use tracing::debug;
 
 use crate::{
-    hp::{
-        self,
-        cfg::{self, DERP_MAGIC_IP},
-        derp::DerpMap,
-        netmap,
-    },
+    endpoint::MagicEndpoint,
+    hp::derp::DerpMap,
     tls::{self, Keypair, PeerId},
 };
 
-/// Create a quinn client endpoint
+/// Connect to a remote endpoint, creating an endpoint on the fly.
 ///
-/// The *bind_addr* is the address that should be bound locally.  Even though this is an
-/// outgoing connection a socket must be bound and this is explicit.  The main choice to
-/// make here is the address family: IPv4 or IPv6.  Otherwise you normally bind to the
-/// `UNSPECIFIED` address on port `0` thus allowing the kernel to do the right thing.
-///
-/// If *peer_id* is present it will verify during the TLS connection setup that the remote
-/// connected to has the required [`PeerId`], otherwise this will connect to any peer.
-///
-/// The *alpn_protocols* are the list of Application-Layer Protocol Neotiation identifiers
-/// you are happy to accept.
-///
-/// If *keylog* is `true` and the KEYLOGFILE environment variable is present it will be
-/// considered a filename to which the TLS pre-master keys are logged.  This can be useful
-/// to be able to decrypt captured traffic for debugging purposes.
-///
-/// Finally the *derp_map* specifies the DERP servers that can be used to establish this
-/// connection.
-pub async fn create_endpoint(
-    bind_addr: SocketAddr,
-    peer_id: PeerId,
-    alpn_protocols: Vec<Vec<u8>>,
-    keylog: bool,
-    derp_map: Option<DerpMap>,
-) -> Result<(quinn::Endpoint, hp::magicsock::Conn)> {
-    let keypair = Keypair::generate();
-
-    let tls_client_config =
-        tls::make_client_config(&keypair, Some(peer_id), alpn_protocols, keylog)?;
-    let mut client_config = quinn::ClientConfig::new(Arc::new(tls_client_config));
-
-    let conn = hp::magicsock::Conn::new(hp::magicsock::Options {
-        port: bind_addr.port(),
-        private_key: keypair.secret().clone().into(),
-        ..Default::default()
-    })
-    .await?;
-    conn.set_derp_map(derp_map).await?;
-
-    let mut endpoint = quinn::Endpoint::new_with_abstract_socket(
-        quinn::EndpointConfig::default(),
-        None,
-        conn.clone(),
-        Arc::new(quinn::TokioRuntime),
-    )?;
-
-    let mut transport_config = quinn::TransportConfig::default();
-    transport_config.keep_alive_interval(Some(Duration::from_secs(1)));
-    client_config.transport_config(Arc::new(transport_config));
-
-    endpoint.set_default_client_config(client_config);
-    Ok((endpoint, conn))
-}
-
-/// Establishes a QUIC connection to the provided peer.
+/// The PeerId and the ALPN protocol are required. If you happen to know dialable addresses of
+/// the remote endpoint, they can be specified and will be added to the endpoint's peer map.
+/// If no addresses are specified, the endpoint will try to dial the peer through the
+/// configured DERP servers.
 pub async fn dial_peer(
```

Comment:
I think it would be nice to just make this a helper function on `MagicEndpoint`, eg `MagicEndpoint::dial_peer(..) -> Self`. That way we can reduce the API surface a little bit more, and all interactions just go through `MagicEndpoint`

Reply:
Yup, then we can remove the client module alltogether. The other function is only used in a single place (RPC client) and has not much to do with iroh_net as its just creating a quinn endpoint, so I'd just move that to where the RPC client is created for now.